desc/compute/bounce_integral.py

```diff
@@ -0,0 +1,1386 @@
+
+        Specify to only return the first ``size`` pairs of bounce points for each
+        pitch along each field line. This is useful if ``size`` is a close upper
+        bound to the actual number of wells. To get an intuition for a good estimate
+        for ``size``, plot the field line with all the bounce points identified
+        by calling this function with ``check=True``.
+        As a reference, there are typically <= 5 wells per toroidal transit.
+    check : bool
+        Flag for debugging.
+    plot : bool
+        Whether to plot some things if check is true. Default is true.
+
+    Returns
+    -------
+    bp1, bp2 : (jnp.ndarray, jnp.ndarray)
+        Shape (P, S, num_wells).
+        The field line-following coordinates of bounce points for a given pitch along
+        a field line. The pairs ``bp1`` and ``bp2`` form left and right integration
+        boundaries, respectively, for the bounce integrals.
+
+        If there were less than ``size`` wells detected along a field line, then
+        the last axis, which enumerates bounce points for  a particular field line
+        and pitch, is padded with zero.
+
+    """
+    B_c, B_z_ra_c, pitch = _check_shape(knots, B_c, B_z_ra_c, pitch)
+    P, S, N, degree = pitch.shape[0], B_c.shape[1], knots.size - 1, B_c.shape[0] - 1
+    # Intersection points in local power basis.
+    intersect = _poly_root(
+        c=B_c,
+        k=jnp.reciprocal(pitch)[..., jnp.newaxis],
+        a_min=jnp.array([0]),
+        a_max=jnp.diff(knots),
+        sort=True,
+        sentinel=-1,
+        distinct=True,
+    )
+    assert intersect.shape == (P, S, N, degree)
+
+    # Reshape so that last axis enumerates intersects of a pitch along a field line.
+    B_z_ra = _poly_val(x=intersect, c=B_z_ra_c[..., jnp.newaxis]).reshape(P, S, -1)
+    # Only consider intersect if it is within knots that bound that polynomial.
+    is_intersect = intersect.reshape(P, S, -1) >= 0
+    # Following discussion on page 3 and 5 of https://doi.org/10.1063/1.873749,
+    # we ignore the bounce points of particles only assigned to a class that are
+    # trapped outside this snapshot of the field line.
+    is_bp1 = (B_z_ra <= 0) & is_intersect
+    is_bp2 = (B_z_ra >= 0) & _fix_inversion(is_intersect, B_z_ra)
+
+    # Transform out of local power basis expansion.
+    intersect = (intersect + knots[:-1, jnp.newaxis]).reshape(P, S, -1)
+    # New versions of jax only like static sentinels.
+    sentinel = -10000000.0  # knots[0] - 1
+    bp1 = _take_mask(intersect, is_bp1, size=num_wells, fill_value=sentinel)
+    bp2 = _take_mask(intersect, is_bp2, size=num_wells, fill_value=sentinel)
+
+    if check:
+        _check_bounce_points(bp1, bp2, sentinel, pitch, knots, B_c, plot, **kwargs)
+
+    mask = (bp1 > sentinel) & (bp2 > sentinel)
+    # Set outside mask to same value so integration is over set of measure zero.
+    bp1 = jnp.where(mask, bp1, 0)
+    bp2 = jnp.where(mask, bp2, 0)
```

Comment:
programming optimization: could use jnp.searchsorted here since bp1 and bp2 are ordered arrays